### PR TITLE
fix(amazonq): update logic for converting gitignore pattern to regex

### PR DESCRIPTION
**Pull request by gandhi-21**
Updating .gitignore pattern to regex conversion to factor for following edge cases:
* `.*` should be matching only files starting with `.` now, instead of all files previously

As part of the changes, added a unit test to verify the correct regex pattern is being created and files are being matched. 
Existing tests which check the file zipping logic are passing as well so no regression should be expected.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

The current implementation to convert `.gitignore` patterns into regex is not following gitignore patterns and resulting in the below issues

* `.*` is converted into `\\..*/.*` which results in all files being selected instead of the intention to only select files that start with a `.`. This is a special case where adding `^` in the front makes sure only files starting with `.` are ignored.

### Changes
* Added special handling for `.*` case
* Added a unit test to check the regex patterns are matching the files as expected from the changes

## Checklist
- [X] My code follows the code style of this project
- [X] I have added tests to cover my changes
- [N/A] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [N/A] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt [hunk 1]
```diff
@@ -161,4 +161,40 @@ class FeatureDevSessionContextTest : FeatureDevTestBase(HeavyJavaCodeInsightTest
...
     }
+
+    @Test
+    fun testConvertGitIgnorePatternToRegex() {
+        val sampleGitIgnorePatterns = listOf(".*", "build/", "*.txt", "**/*.png")
+        val sampleFileNames = listOf(
+            ".gitignore/",
+            ".env/",
+            "file.txt/",
+            ".git/config/",
```
Comment (juanbenj) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:172]: I think `.*` matches both files and dirs that begin with a dot

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt [hunk 2]
```diff
@@ -161,4 +161,40 @@ class FeatureDevSessionContextTest : FeatureDevTestBase(HeavyJavaCodeInsightTest
...
+            "build/output.jar/",
+            "builds/",
+            "mybuild/",
+            "build.json/",
+            "log.txt/",
+            "file.txt.json/",
+            "file.png/",
+            "src/file.png/"
+        )
+        val expectedFilesToMatch = listOf(".gitignore/", ".env/", "file.txt/", "build/", "build/output.jar/", "log.txt/", "file.png/", "src/file.png/")
```
Comment (juanbenj) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:184]: NIT: I would move this directly above the assertion -- it took me a bit to understand what this meant in the context of the test

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt [hunk 3]
```diff
@@ -161,4 +161,40 @@ class FeatureDevSessionContextTest : FeatureDevTestBase(HeavyJavaCodeInsightTest
...
+        val patterns = sampleGitIgnorePatterns.map { pattern -> featureDevSessionContext.convertGitIgnorePatternToRegex(pattern).toRegex() }
+
+        for (sampleFileName in sampleFileNames) {
+            for (pattern in patterns) {
+                if (pattern.matches(sampleFileName)) {
+                    matchedFiles.add(sampleFileName)
+                    break
+                }
+            }
+        }
```
Comment (juanbenj) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:196]: Optional: You could do something like:

```kotlin
val matchedFiles = samplefileNames.filter { fileName -> patterns.find { pattern -> pattern.matches(fileName)}}
```

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt [hunk 4]
```diff
@@ -279,10 +279,25 @@ class FeatureDevSessionContext(val project: Project, val maxProjectSizeBytes: Lo
...
+        // Special case for ".*" to match only dotfiles
+        if (pattern.equals(".*")) {
+            return "^\\.[^/]+/?$"
+        }
+
+        return pattern
+            .replace(".", "\\.")
+            .replace("**/", "(.*/)?")
+            .replace("**", ".*")
+            .replace("*", "[^/]*")
```
Comment (juanbenj) [inline plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt:292]: Will this line start replacing the changes from L289 & L290?

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt (hunk 3 above)
Comment (gandhi-21) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:196]: makes sense, updating

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt [hunk 5]
```diff
@@ -161,4 +161,40 @@ class FeatureDevSessionContextTest : FeatureDevTestBase(HeavyJavaCodeInsightTest
...
+
+    @Test
+    fun testConvertGitIgnorePatternToRegex() {
+        val sampleGitIgnorePatterns = listOf(".*", "build/", "*.txt", "**/*.png")
+        val sampleFileNames = listOf(
+            ".gitignore/",
+            ".env/",
+            "file.txt/",
+            ".git/config/",
+            "src/file.txt/",
```
[the remark this answers is not in the document]
Comment (gandhi-21) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:173]: you're right, just tried and checked, updating the logic

File: plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt (hunk 4 above)
Comment (gandhi-21) [reply plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt]: changed back to original implementation and extracted the special case, the rest changes including this were based on incorrect assumption of gitignore pattern

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt (hunk 2 above)
Comment (gandhi-21) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt]: updated

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt (hunk 1 above)
Comment (gandhi-21) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:172]: also updated the test for this

**Review by juanbenj — APPROVED** [review]

**Review by neilk-aws — APPROVED** [review]

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt [hunk 6]
```diff
@@ -161,4 +161,40 @@ class FeatureDevSessionContextTest : FeatureDevTestBase(HeavyJavaCodeInsightTest
...
+            "build/",
+            "build/output.jar/",
+            "builds/",
+            "mybuild/",
+            "build.json/",
+            "log.txt/",
+            "file.txt.json/",
```
[the remark this answers is not in the document]
Comment (gandhi-21) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/FeatureDevSessionContextTest.kt:180]: there's a trailing / added when matching files as per the comment 
https://github.com/aws/aws-toolkit-jetbrains/blob/main/plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt#L285

**Review by rli — APPROVED** [review]
